File: backend/app/routers/health.py
from fastapi import APIRouter, Depends
from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession
import time

from app.database import get_db

router = APIRouter(prefix="/health", tags=["health"])

START_TIME = time.time()


@router.get("")
async def health(db: AsyncSession = Depends(get_db)):
    db_status = "ok"
    db_latency = None

    try:
        t = time.time()
        await db.execute(text("SELECT 1"))
        db_latency = round((time.time() - t) * 1000, 2)
    except Exception as e:
        db_status = f"error: {str(e)}"

    uptime = round(time.time() - START_TIME)

    return {
        "status": "ok" if db_status == "ok" else "degraded",
        "uptime_seconds": uptime,
        "services": {
            "api": {"status": "ok"},
            "database": {
                "status": db_status,
                "latency_ms": db_latency,
            },
        },
    }

File: container-engine/src/handlers/health.rs
use std::sync::{Arc, OnceLock};
use std::time::Instant;

use axum::{extract::State, routing::get, Json, Router};

use crate::state::AppState;

pub fn router() -> Router<Arc<AppState>> {
    Router::new().route("/", get(health))
}

fn start_time() -> Instant {
    static START: OnceLock<Instant> = OnceLock::new();
    *START.get_or_init(Instant::now)
}

async fn health(State(state): State<Arc<AppState>>) -> Json<serde_json::Value> {
    let started = start_time();
    let db_started = Instant::now();
    let db_result = sqlx::query("SELECT 1").fetch_one(&state.db).await;
    let db_latency = ((db_started.elapsed().as_secs_f64()) * 1000.0 * 100.0).round() / 100.0;

    let (db_status, db_latency_ms) = match db_result {
        Ok(_) => ("ok".to_string(), Some(db_latency)),
        Err(e) => (format!("error: {e}"), None),
    };

    let uptime_seconds = started.elapsed().as_secs();
    let overall = if db_status == "ok" { "ok" } else { "degraded" };

    Json(serde_json::json!({
        "status": overall,
        "uptime_seconds": uptime_seconds,
        "services": {
            "api": { "status": "ok" },
            "database": {
                "status": db_status,
                "latency_ms": db_latency_ms,
            }
        }
    }))
}
